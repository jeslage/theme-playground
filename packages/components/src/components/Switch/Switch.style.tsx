import styled from 'styled-components';

interface SwitchProps {
  readonly active: boolean;
}

const StyledSwitch = styled.div<SwitchProps>`
  display: flex;
  align-items: center;
  margin: 1.5em 0;
  width: 100%;

  button {
    position: relative;
    outline: none;
    margin: 0;
    padding: 0;
    appearance: none;
    border: 1px solid ${({ theme }) => theme.text};
    width: 60px;
    height: 30px;
    border-radius: 30px;
    cursor: pointer;
    background: transparent;

    span {
      position: absolute;
      top: 4px;
      left: 4px;
      display: block;
      width: 20px;
      height: 20px;
      border-radius: 20px;
      background: ${({ theme }) => theme.text};
      transform: translateX(${props => (props.active ? '30px' : '0')});
      will-change: transform;
      transition: transform 0.2s ease-in-out;

      svg {
        fill: ${props => props.theme.backgroundLight};
        fill-opacity: 0;
        stroke: ${props => props.theme.backgroundLight};
        stroke-width: 3;
        stroke-linecap: round;
        stroke-linejoin: round;
      }
    }
  }
`;

export default StyledSwitch;
