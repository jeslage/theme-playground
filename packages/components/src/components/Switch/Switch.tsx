import * as React from 'react';

import StyledSwitch from './Switch.style';
import Label from '../Label/Label';

export interface Props {
  iconBefore?: JSX.Element;
  label?: string;
  description?: string;
  onChange: (val: boolean) => void;
  value: boolean;
}

const Switch: React.FC<Props> = ({
  iconBefore,
  label,
  onChange,
  description,
  value
}) => (
  <StyledSwitch active={value}>
    <Label iconBefore={iconBefore} label={label} description={description} />

    <button
      onClick={() => {
        if (onChange) onChange(!value);
      }}
      type="button"
      role="switch"
      aria-checked={value}
    >
      <span>
        {!value ? (
          <svg
            viewBox="0 0 52 52"
            focusable="false"
            aria-hidden="true"
            role="presentation"
          >
            <path d="m19.1 19.1 l14 14 m 0 -14 l -14 14" />
          </svg>
        ) : (
          <svg
            viewBox="0 0 52 52"
            focusable="false"
            aria-hidden="true"
            role="presentation"
          >
            <path d="m19.1 25.2 4.7 6.2 12.1-11.2" />
          </svg>
        )}
      </span>
    </button>
  </StyledSwitch>
);

export default Switch;
