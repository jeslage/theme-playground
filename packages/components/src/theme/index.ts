export const darkTheme = {
  base: 'dark',
  backgroundDark: '#2f2f2f',
  backgroundLight: '#333',
  border: '#999999',
  text: '#FFFFFF'
};

export const lightTheme = {
  base: 'light',
  backgroundDark: '#f6f9fc',
  backgroundLight: '#fff',
  border: '#666666',
  text: '#333333'
};

export const getUiTheme = theme => {
  if (theme === undefined) return darkTheme;

  if (typeof theme === 'string') {
    if (theme === 'dark') {
      return darkTheme;
    } else if (theme === 'light') {
      return lightTheme;
    }

    throw Error(
      `Theme "${theme}" is not available. Please use "light" or "dark", or provide a custom theme object.`
    );
  }

  if (theme.base && theme.base === 'light') {
    return {
      ...lightTheme,
      ...theme
    };
  }

  return {
    ...darkTheme,
    ...theme
  };
};
