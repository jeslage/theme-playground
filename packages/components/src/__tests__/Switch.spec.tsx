import 'jest';
import * as React from 'react';
import { shallow } from 'enzyme';
import Switch from '../components/Switch/Switch';
import StyledSwitch from '../components/Switch/Switch.style';
import { darkTheme } from '../theme';

describe('Switch', () => {
  it('Enabled Switch should render correctly', () => {
    const component = shallow(
      <Switch
        onChange={value => console.log(value)}
        label="Label"
        description="Testbeschreibung"
        value={true}
        iconBefore={
          <svg height="24" viewBox="0 0 24 24" width="24">
            <rect height="2" rx="1" width="12" x="6" y="11" />
          </svg>
        }
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('Disabled Switch should render correctly', () => {
    const component = shallow(
      <Switch
        onChange={value => console.log(value)}
        label="Label"
        description="Testbeschreibung"
        value={false}
        iconBefore={
          <svg height="24" viewBox="0 0 24 24" width="24">
            <rect height="2" rx="1" width="12" x="6" y="11" />
          </svg>
        }
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('StyledSwitch should render correctly', () => {
    const component = shallow(<StyledSwitch theme={{ colors: darkTheme }} />);

    expect(component).toMatchSnapshot();
  });
});
