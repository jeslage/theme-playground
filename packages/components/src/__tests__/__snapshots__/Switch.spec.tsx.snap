// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Disabled Switch should render correctly 1`] = `
<styled.div
  active={false}
>
  <Label
    description="Testbeschreibung"
    iconBefore={
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <rect
          height="2"
          rx="1"
          width="12"
          x="6"
          y="11"
        />
      </svg>
    }
    label="Label"
  />
  <button
    aria-checked={false}
    onClick={[Function]}
    role="switch"
    type="button"
  >
    <span>
      <svg
        aria-hidden="true"
        focusable="false"
        role="presentation"
        viewBox="0 0 52 52"
      >
        <path
          d="m19.1 19.1 l14 14 m 0 -14 l -14 14"
        />
      </svg>
    </span>
  </button>
</styled.div>
`;

exports[`Switch Enabled Switch should render correctly 1`] = `
<styled.div
  active={true}
>
  <Label
    description="Testbeschreibung"
    iconBefore={
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <rect
          height="2"
          rx="1"
          width="12"
          x="6"
          y="11"
        />
      </svg>
    }
    label="Label"
  />
  <button
    aria-checked={true}
    onClick={[Function]}
    role="switch"
    type="button"
  >
    <span>
      <svg
        aria-hidden="true"
        focusable="false"
        role="presentation"
        viewBox="0 0 52 52"
      >
        <path
          d="m19.1 25.2 4.7 6.2 12.1-11.2"
        />
      </svg>
    </span>
  </button>
</styled.div>
`;

exports[`Switch StyledSwitch should render correctly 1`] = `
<div
  className="sc-AxjAm gdUeta"
/>
`;
